arbnode/batch_poster.go

```diff
@@ -1087,13 +1087,74 @@ func (s *batchSegments) CloseAndGetBytes() ([]byte, error) {
 	return fullMsg, nil
 }
 
-// Make the batch wait for validation Add this so we don't need to export the structs state to set it as we shouldn't need to set it to false again.
-func (s *batchSegments) SetWaitingForValidation() {
-	if !s.isWaitingForEspressoValidation {
-		log.Info("Set current batch segments to waiting for validation")
-		s.isWaitingForEspressoValidation = true
+// resetStreamerToParentChainOrConfigHotshotBlock:
+//
+//	This function reads state from external sources and resets the espresso streamer to start producing
+//	messages from hotshot based on the source of truth on the parent chain
+//
+// Parameters:
+//
+//	messageCount: an arbutil.MessageIndex representing the current message position that the espresso streamer should be searching for.
+//	This value should be obtained from the parent chains sequencer inbox as a source of truth, callers providing incorrect values will
+//	reset the streamer to a state that will now allow the batch poster to post batches.
+//
+// Returns:
+//
+//	None.
+func (b *BatchPoster) resetStreamerToParentChainOrConfigHotshotBlock(messageCount arbutil.MessageIndex, ctx context.Context) {
+	hotshotBlock := b.fetchHotshotBlockFromLastCheckpoint(ctx)
+	if hotshotBlock == 0 {
+		// if there hasn't been a batch posted, or we encountered an error, start reading from the configured hotshot block number.
+		hotshotBlock = b.config().HotShotBlock
+	}
+	b.espressoStreamer.Reset(uint64(messageCount), hotshotBlock)
+}
+
+// fetchHotshotBlockFromLastCheckpoint:
+// This function uses the sequencer inbox bridgegen contract to filter for logs related to the TEESignatureVerified events
+// If any of these events are encountered, it checks the log iterator for the data about this event.
+// Return:
+// returns the Hotshot height of the last event in the iterator returned from FilterTEESignatureVerified()
+// representing the most recently emitted hotshotblock height. Any errors encountered will result in 0 being returned.
+func (b *BatchPoster) fetchHotshotBlockFromLastCheckpoint(ctx context.Context) uint64 {
+	pollingStep := b.config().EspressoEventPollingStep
+	header, err := b.l1Reader.LastHeader(ctx)
+	if err != nil {
+		log.Error("Failed to fetch last header from parent chain")
+		return 0
+	}
+
+	var logIterator *bridgegen.SequencerInboxTEESignatureVerifiedIterator
+	for i := header.Number.Uint64(); i >= b.config().HotShotGenesisBlock; i -= pollingStep {
```

Comment:
nit: instead of doing it iteratively, we could have just made this a recursive function. It would have looked a bit cleaner